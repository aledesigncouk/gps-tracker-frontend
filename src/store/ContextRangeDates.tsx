import React, { createContext, useState, ReactNode } from 'react';

interface RangeDatesContextProps {
    startDate: Date | null;
    endDate: Date | null;
    setStartDate: (date: Date | null) => void;
    setEndDate: (date: Date | null) => void;
}

const RangeDatesContext = createContext<RangeDatesContextProps | undefined>(undefined);

export const RangeDatesContextStore: React.FC<{ children: ReactNode }> = ({ children }) => {
    const [startDate, setStartDate] = useState<Date | null>(null);
    const [endDate, setEndDate] = useState<Date | null>(null);

    return (
        <RangeDatesContext.Provider value={{
            startDate: startDate,
            setStartDate: setStartDate,
            endDate: endDate,
            setEndDate: setEndDate,
        }}>
            {children}
        </RangeDatesContext.Provider>
    );
};

export const useRangeDatesStore = () => {
    const context = React.useContext(RangeDatesContext);
    if (context === undefined) {
        throw new Error('useRangeDatesStore must be used within a RangeDatesProvider');
    }
    return context;
};

export default RangeDatesContextStore;