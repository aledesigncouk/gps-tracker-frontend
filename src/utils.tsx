export const getTrackByRange = async (start, end) => {
  const response = await fetch(
    `/api/TrackByRange?start=${encodeURIComponent(
      start,
    )}?end=${encodeURIComponent(end)}`,
  );

  const result = await response.json();
  return result;
};

export const getYears = async () => {
  const response = await fetch("/api/GetYears");

  const result = await response.json();
  return result;
};

export const setRangeByYear = (year: string): {startDate: string, endDate: string} => {
  const startDate = `${year}-01-01 00:00:00`;
  const endDate = `${year}-12-31 00:00:00`;

  return { startDate, endDate };
}
