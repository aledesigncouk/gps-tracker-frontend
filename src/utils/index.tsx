// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import React, { useEffect, useState } from 'react'

// fetch ALL data, filters applied on frontend
export const fetchData = () => {
	const [allData, setData] = useState([])
	useEffect(() => {
		fetch('https://www.yoroxid.com/nautilus/items/read.php')
			.then((response) => response.json())
			.then((data) => {
				console.log(data)
				setData(data)
			})
			.catch((err) => {
				console.log(err.message)
			})
	}, [])

	console.log('data:', allData)
}

/* get tracks year
 */
export const getTrackByYear = (dataset) => {
	return 'track from range'
}

/* get tracks by data range
 */
export const getTrackByRange = () => {
	return 'track from range'
}

/* convert data from database format to geojson

filtro a monte (nella richiesta)
****** INPUT
{
  "points": [
    {
      "id": 1,
      "datatime": "2014-06-01 01:12:26", // properties: { year: '2022' },
      "created": "2014-06-01 01:12:26",

	  // return from server only the list of coords
      "lat": "52.660900",
      "lon": "-2.482342" // geometry: { type: 'LineString', coordinates: [...]
    },
    
  ]
}

for all 
	datatime YYYY-MM-DD => property 

***** OUTPUT
{
	type: 'FeatureCollection',
	features: [
		{
			type: 'Feature',
			properties: { year: '2022' }, // do not set properties for now
			geometry: {
				type: 'LineString',
				coordinates: [
					[52.69204808652421, -2.5520720727985235],
					[52.688930911955794, -2.5304721852188834],
					[52.69080124340172, -2.5167579708822814],
					[52.68809962781904, -2.4739010510800767],
					[52.682903743411345, -2.4475011884826188],
					[52.67937018889114, -2.41252994192439],
					[52.67833085374511, -2.3799586828745873],
					[52.67562846663057, -2.354587386351966],
					[52.66855989455152, -2.329558945188154]
				]
			}
		}, {...}
	]
}
 */
export const geoJsonConverter = (input) => {}
