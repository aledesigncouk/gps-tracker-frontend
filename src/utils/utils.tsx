type Coordinate = [number, number];

interface Geometry {
  type: "LineString";
  coordinates: Coordinate[];
}

interface Properties {
  year: number;
}

interface Track {
  type: "Feature";
  properties: Properties;
  geometry: Geometry;
}

export const getTrackByRange = async (
  start: string,
  end: string
): Promise<Track> => {
  const response = await fetch(`/api/TrackByRange?start=${start}&end=${end}`);

  const data = await response.json();
  return data;
};

export const getYears = async (): Promise<string[]> => {
  const response = await fetch("/api/GetYears");

  const data = await response.json();
  return data;
};

// format the dates to be MySQL friendly
export const formatDate = (date: Date): string => {
  const yyyy = date.getFullYear();
  const mm = String(date.getMonth() + 1).padStart(2, "0");
  const dd = String(date.getDate()).padStart(2, "0");

  return `${yyyy}-${mm}-${dd}`;
};

export const setRangeByYear = (
  year: string
): { startDate: string; endDate: string } => {
  const startDate = `${year}-01-01`;
  const endDate = `${year}-12-31`;

  return { startDate, endDate };
};

export const validateTrackData = (data: Track): boolean => {
  if (
    data &&
    data.type === "Feature" &&
    data.properties &&
    typeof data.properties.year === "number" &&
    data.geometry &&
    data.geometry.type === "LineString" &&
    Array.isArray(data.geometry.coordinates) &&
    data.geometry.coordinates.every(
      (coord) =>
        Array.isArray(coord) &&
        coord.length === 2 &&
        coord.every((val) => typeof val === "number")
    )
  ) {
    return true;
  }
  return false;
};

const validateYearsData = (data: string[]): boolean => {
  if (Array.isArray(data) && data.every((year) => typeof year === "string")) {
    return true;
  }
  return false;
};
