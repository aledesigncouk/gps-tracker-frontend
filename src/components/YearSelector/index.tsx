import { useStore } from "@store/ContextStore";
import { useState, useEffect } from "react";
import { getYears } from "src/utils";

interface YearSelectorProps {
  // onSelect: (value: string) => void;
}

const YearSelector: React.FC<YearSelectorProps> = () => {
  const {selectedYear, setSelectedYear} = useStore(); // selected year
  const [years, setYears] = useState<string[]>([]); // list of available years

  useEffect(() => {
    const fetchYears = async () => {
      try {
        const result = await getYears(); 
        setYears(Array.isArray(result) ? result : []); 
      } catch (error) {
        console.error("Failed to fetch years:", error); 
        setYears([]);
      }
    };

    fetchYears();
  }, []); 

  const handleChange = (event) => {
    const value = event.target.value;
    setSelectedYear(value);
  };

  return (
    <div>
      <label htmlFor="dropdown">Choose a year:</label>
      <select
        id="dropdown"
        value={selectedYear}
        onChange={handleChange}
        style={{ marginLeft: "10px", padding: "5px" }}
      >
         <option value="empty">Years</option>
        {years.map((option, index) => (
          <option key={index} value={option}>
            {option}
          </option>
        ))}
      </select>
    </div>
  );
};

export default YearSelector;
