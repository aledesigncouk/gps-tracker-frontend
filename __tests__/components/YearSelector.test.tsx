import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import YearSelector from '@components/YearSelector';
import { useRangeDatesStore } from '@store/ContextRangeDates';
import { getYears, setRangeByYear } from '@utils/utils';

jest.mock('@store/ContextRangeDates', () => ({
  useRangeDatesStore: jest.fn(),
}));

jest.mock('@utils/utils', () => ({
  getYears: jest.fn(),
  setRangeByYear: jest.fn(),
}));

describe('YearSelector component', () => {
  const setStartDate = jest.fn();
  const setEndDate = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    useRangeDatesStore.mockReturnValue({
      setStartDate,
      setEndDate,
    });

    getYears.mockResolvedValue(['2021', '2022', '2023', '2024']);
  });

  it('renders the YearSelector component', async () => {
    render(<YearSelector />);

    await waitFor(() => {
      expect(screen.getByTestId('year-selector')).toBeInTheDocument;
    });
  });

  it('renders the dropdown with available years', async () => {
    const years = ['2021', '2022', '2023', '2024'];

    render(<YearSelector />);

    await waitFor(() => {
      const dropdownToggle = screen.getByRole('button', { name: /select year/i });
      fireEvent.click(dropdownToggle);

      years.forEach(year => {
        expect(screen.getByText(year)).toBeInTheDocument();
      });
    });
  });

  it('fires setStartDate and setEndDate functions when a year is selected', async () => {
    setRangeByYear.mockReturnValue({
      startDate: '2023-01-01',
      endDate: '2023-12-31',
    });

    render(<YearSelector />);

    await waitFor(() => {
      const dropdownToggle = screen.getByRole('button', { name: /select year/i });
      fireEvent.click(dropdownToggle);

      const yearItem = screen.getByText('2023');
      fireEvent.click(yearItem);

      expect(setStartDate).toHaveBeenCalledWith(new Date('2023-01-01'));
      expect(setEndDate).toHaveBeenCalledWith(new Date('2023-12-31'));
    });
  });

  it('does not call setStartDate or setEndDate if no year is selected', async () => {

    render(<YearSelector />);

    await waitFor(() => {
      const dropdownToggle = screen.getByRole('button', { name: /select year/i });
      fireEvent.click(dropdownToggle);
      fireEvent.click(document.body);

      expect(setStartDate).not.toHaveBeenCalled();
      expect(setEndDate).not.toHaveBeenCalled();
    });
  });

  it('handles error when getYears fails', async () => {
    getYears.mockRejectedValueOnce(new Error('Failed to fetch years'));

    render(<YearSelector />);

    await waitFor(() => {
      const dropdownToggle = screen.getByRole('button', { name: /select year/i });
      expect(dropdownToggle).toHaveTextContent('Select year');
    });
  });

  it('fires setStartDate and setEndDate with expected values when a different year is selected', async () => {
    setRangeByYear.mockReturnValueOnce({
      startDate: '2021-01-01',
      endDate: '2021-12-31',
    });
    setRangeByYear.mockReturnValueOnce({
      startDate: '2022-01-01',
      endDate: '2022-12-31',
    });

    render(<YearSelector />);

    await waitFor(() => {
      const dropdownToggle = screen.getByRole('button', { name: /select year/i });
      fireEvent.click(dropdownToggle);

      const year2021 = screen.getByText('2021');
      fireEvent.click(year2021);

      expect(setStartDate).toHaveBeenCalledWith(new Date('2021-01-01'));
      expect(setEndDate).toHaveBeenCalledWith(new Date('2021-12-31'));

      const year2022 = screen.getByText('2022');
      fireEvent.click(year2022);

      expect(setStartDate).toHaveBeenCalledWith(new Date('2022-01-01'));
      expect(setEndDate).toHaveBeenCalledWith(new Date('2022-12-31'));

    })

  });
});

describe('YearSelector component', () => {
  it('match the snapshot', async () => {
    await waitFor(() => {
      const { asFragment } = render(<YearSelector />);
      expect(asFragment()).toMatchSnapshot();
    })
  });
});
