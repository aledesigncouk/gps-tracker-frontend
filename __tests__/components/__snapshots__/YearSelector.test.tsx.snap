// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearSelector component match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ms-auto vstack"
    data-testid="year-selector"
  >
    <label
      class=""
      for="dropdown"
    >
      Choose a year:
    </label>
    <div
      class="dropdown"
    >
      <button
        aria-expanded="false"
        class="dropdown-toggle btn btn-success"
        data-testid="year-dropdown"
        id="dropdown-basic"
        type="button"
      >
        Select year
      </button>
    </div>
  </div>
</DocumentFragment>
`;
